libgadget/uvbg.c

```diff
@@ -0,0 +1,971 @@
+        break;
+    case INDEX_REAL:
+        ind = k + dim * (j + dim * i);
+        break;
+    case INDEX_COMPLEX_HERM:
+        ind = k + (dim / 2 + 1) * (j + dim * i);
+        break;
+    default:
+        endrun(1, "Unknown indexing type in `grid_index`.");
+        break;
+    }
+
+    return ind;
+}
+
+
+static void populate_grids()
+{
+    // TODO(smutch): Is this stored somewhere globally?
+    int uvbg_dim = All.UVBGdim;
+    int nranks = -1, this_rank = -1;
+    MPI_Comm_size(MPI_COMM_WORLD, &nranks);
+    MPI_Comm_rank(MPI_COMM_WORLD, &this_rank);
+    ptrdiff_t *slab_nix = UVBGgrids.slab_nix;
+    ptrdiff_t *slab_ix_start = UVBGgrids.slab_ix_start;
+
+    // create buffers on each rank which is as large as the largest LOGICAL allocation on any single rank
+    int buffer_size = 0;
+    for (int ii = 0; ii < nranks; ii++)
+        if (slab_nix[ii] > buffer_size)
+            buffer_size = (int)slab_nix[ii];
+
+    buffer_size *= uvbg_dim * uvbg_dim;
+    float *buffer_mass = fftwf_alloc_real((size_t)buffer_size);
+    float *buffer_stars_slab = fftwf_alloc_real((size_t)buffer_size);
+    float *buffer_sfr = fftwf_alloc_real((size_t)buffer_size);
+
+    //RegionInd no longer global, allocate array for slab decomposition (Jdavies)
+    int *UVRegionInd = mymalloc("UVRegionInd",sizeof(int) * PartManager->NumPart);
+
+    // I am going to reuse the RegionInd member which is from petapm.  I
+    // *think* this is ok as we are doing this after the grav calculations.
+    // This is a potentially stupid way to do things anyway and will most
+    // definitely need to be changed! There is no way we should have to search
+    // all of the particles to find out what slab they sit on, and then loop
+    // through all particles again n_slab times!
+    double box_size = All.BoxSize;
+
+    #pragma omp parallel for
+    for(int ii = 0; ii < PartManager->NumPart; ii++) {
+        if((!P[ii].IsGarbage) && (!P[ii].Swallowed) && (P[ii].Type < 5)) {
+            ptrdiff_t ix = pos_to_ngp(P[ii].Pos[0], box_size, uvbg_dim);
+            UVRegionInd[ii] = searchsorted(&ix, slab_ix_start, nranks, sizeof(ptrdiff_t), compare_ptrdiff, -1, -1);
+        } else {
+            UVRegionInd[ii] = -1;
+        }
+    }
+
+
+    for (int i_r = 0; i_r < nranks; i_r++) {
+
+        const int ix_start = slab_ix_start[i_r];
+        const int nix = slab_nix[i_r];
+
+        // init the buffers
+        for (int ii = 0; ii < buffer_size; ii++) {
+            buffer_mass[ii] = (float)0.;
+            buffer_stars_slab[ii] = (float)0.;
+            buffer_sfr[ii] = (float)0.;
+        }
+
+
+        // fill the local buffer for this slab
+        // TODO(smutch): This should become CIC
+        unsigned int count_mass = 0;
+        #pragma omp parallel for reduction(+:count_mass)
+        for(int ii = 0; ii < PartManager->NumPart; ii++) {
+            if(UVRegionInd[ii] == i_r) {
+                int ix = pos_to_ngp(P[ii].Pos[0], box_size, uvbg_dim) - ix_start;
```

Comment:
I think you need CurrentParticleOffset here

Reply:
I've added the offset to pos_to_ngp